### PR TITLE
[Merged by Bors] - feat: Rademacher theorem

### PR DESCRIPTION
**Pull request by sgouezel**
We prove Rademacher theorem, stating that a Lipschitz function between finite-dimensional real vector spaces is differentiable almost everywhere.

---

- [x] depends on: #6903 
- [x] depends on: #6979 
- [x] depends on: #7004 
- [x] depends on: #7005
- [x] depends on: #7006
- [x] depends on: #7038 
- [x] depends on: #7055
- [x] depends on: #7252

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Analysis/Calculus/Rademacher.lean [hunk 1]
```diff
@@ -0,0 +1,404 @@
...
+    apply LipschitzWith.ae_differentiableAt_real (C := C)
+    intro z z'
+    convert hf (x, z) (x, z')
+    simp [Prod.edist_eq]
+  filter_upwards [this] with y hy
+  have h'y : DifferentiableAt ℝ (fun z ↦ f (x, z)) (y + 0) := by simpa using hy
+  have : DifferentiableAt ℝ (fun t ↦ y + t) 0 :=
+    (differentiable_id.const_add _).differentiableAt
+  simpa only [LineDifferentiableAt, Prod.smul_mk, smul_zero, smul_eq_mul, mul_one, Prod.mk_add_mk,
+    add_zero] using h'y.comp 0 this
```
Comment (urkud) [inline Mathlib/Analysis/Calculus/Rademacher.lean:74]: Optional golf:
```suggestion
  have : ∀ᵐ (y : ℝ), DifferentiableAt ℝ (fun z ↦ f (x, z)) y :=
    (hf.comp (LipschitzWith.prod_mk_left x)).ae_differentiableAt_real
  filter_upwards [this] with y hy
  have h'y : DifferentiableAt ℝ (fun z ↦ f (x, z)) (y + 0) := by simpa using hy
  have : DifferentiableAt ℝ (fun t ↦ y + t) 0 := differentiableAt_id.const_add _
  simpa [LineDifferentiableAt] using h'y.comp 0 this
```

File: Mathlib/Analysis/Calculus/Rademacher.lean [hunk 2]
```diff
@@ -0,0 +1,404 @@
...
+
+theorem ae_lineDifferentiableAt (hf : LipschitzWith C f) (v : E) :
+    ∀ᵐ p ∂μ, LineDifferentiableAt ℝ f p v := by
```
Comment (urkud) [inline Mathlib/Analysis/Calculus/Rademacher.lean:77]: Does it make sense to have a generic theorem "if a predicate is true a.e. on every `k`-dimensional (`1`-dimensional?) subspace, then it's true a.e.-everywhere w.r.t. a Haar measure"?

**Comment by urkud** [comment]
Thanks!
bors d+

File: Mathlib/Analysis/Calculus/Rademacher.lean [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+import Mathlib.Analysis.BoundedVariation
+import Mathlib.Analysis.NormedSpace.HahnBanach.SeparatingDual
+import Mathlib.MeasureTheory.Group.Integral
+import Mathlib.Analysis.Distribution.AEEqOfIntegralContDiff
+
+/-!
+# Rademacher's theorem: a Lipschitz function is differentiable almost everywhere
+
+This file proves Rademacher's theorem: a Lipschitz function between finite-dimensional real vector
+spaces is differentiable almost everywhere with respect to Lebesgue measure. This is the content
```
Comment (jcommelin) [inline Mathlib/Analysis/Calculus/Rademacher.lean:18]:
```suggestion
spaces is differentiable almost everywhere with respect to the Lebesgue measure. This is the content
```

File: Mathlib/Analysis/Calculus/Rademacher.lean (hunk 2 above)
Comment (sgouezel) [reply Mathlib/Analysis/Calculus/Rademacher.lean:72]: See #7252

**Comment by sgouezel** [comment]
I have streamlined the Fubini argument, by using instead a general argument done in #7252. It's not clear to me that the `bors d+` by @urkud still applies, so I am putting this back to `awaiting-review`.

File: Mathlib/Analysis/Calculus/Rademacher.lean [hunk 4]
```diff
@@ -0,0 +1,377 @@
...
+# Rademacher's theorem: a Lipschitz function is differentiable almost everywhere
+
+This file proves Rademacher's theorem: a Lipschitz function between finite-dimensional real vector
+spaces is differentiable almost everywhere with respect to the Lebesgue measure. This is the content
+of `LipschitzWith.ae_differentiableAt`. Versions for functions which are Lipschitz on sets are also
+given (see `LipschitzOnWith.ae_differentiableWithinAt`).
+
+## Implementation
+
+There are many proofs of Rademacher's theorem. We follow the one by Morrey, which is not the most
```
Comment (RemyDegenne) [inline Mathlib/Analysis/Calculus/Rademacher.lean:24]: Can you add a reference to a paper or a book for the proof by Morrey?

File: Mathlib/Analysis/Calculus/Rademacher.lean [hunk 5]
```diff
@@ -0,0 +1,377 @@
...
+      _ = (C * ‖v‖) *‖g x‖ := by field_simp [norm_smul, abs_of_nonneg t_pos.le]; ring
+      _ = K.indicator (fun x ↦ (C * ‖v‖) * ‖g x‖) x := by rw [indicator_of_mem hx]
+    · have A : f x = 0 := by
+        rw [← Function.nmem_support]
+        contrapose! hx
+        exact self_subset_cthickening _ (subset_tsupport _ hx)
+      have B : f (x + t • v) = 0 := by
+        rw [← Function.nmem_support]
+        contrapose! hx
+        apply mem_cthickening_of_dist_le _ _  (‖v‖) (tsupport f) (subset_tsupport _ hx)
```
Comment (RemyDegenne) [inline Mathlib/Analysis/Calculus/Rademacher.lean:146]:
```suggestion
        apply mem_cthickening_of_dist_le _ _ (‖v‖) (tsupport f) (subset_tsupport _ hx)
```

File: Mathlib/Analysis/Calculus/Rademacher.lean [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+    exact hasLineDerivAt_iff_isLittleO_nhds_zero.1 (hL v (hqs hv))
+  obtain ⟨r, r_pos, hr⟩ : ∃ (r : ℝ), 0 < r ∧ ∀ (t : ℝ), ‖t‖ < r →
+      ∀ v ∈ q, ‖f (x + t • v) - f x - t • L v‖ ≤ δ * ‖t‖ := by
+    rcases Metric.mem_nhds_iff.1 I with ⟨r, r_pos, hr⟩
+    exact ⟨r, r_pos, fun t ht v hv ↦ hr (mem_ball_zero_iff.2 ht) v hv⟩
+  apply Metric.mem_nhds_iff.2 ⟨r, r_pos, fun v hv ↦ ?_⟩
+  rcases eq_or_ne v 0 with rfl|v_ne
+  · simp
+  obtain ⟨w, ρ, w_mem, hvw, hρ⟩ : ∃ w ρ, w ∈ sphere 0 1 ∧ v = ρ • w ∧ ρ = ‖v‖ := by
+    refine ⟨‖v‖⁻¹ •  v, ‖v‖, by simp [norm_smul, inv_mul_cancel (norm_ne_zero_iff.2 v_ne)], ?_, rfl⟩
```
Comment (RemyDegenne) [inline Mathlib/Analysis/Calculus/Rademacher.lean:279]:
```suggestion
    refine ⟨‖v‖⁻¹ • v, ‖v‖, by simp [norm_smul, inv_mul_cancel (norm_ne_zero_iff.2 v_ne)], ?_, rfl⟩
```

**Comment by RemyDegenne** [comment]
Thanks!
bors r+

**Comment by ADedecker** [comment]
bors retry

**Pull request closed without merging**